presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -357,57 +344,30 @@ public void moreMemoryAvailable()
         memoryFuture.get().set(null);
     }
 
-    public void transferMemoryToTaskContext(long taskBytes)
-    {
-        long bytes = memoryReservation.getAndSet(0);
-        driverContext.transferMemoryToTaskContext(bytes);
-
-        TaskContext taskContext = driverContext.getPipelineContext().getTaskContext();
-        if (taskBytes > bytes) {
-            try {
-                taskContext.reserveMemory(taskBytes - bytes);
-            }
-            catch (ExceededMemoryLimitException e) {
-                taskContext.freeMemory(bytes);
-                throw e;
-            }
-        }
-        else {
-            taskContext.freeMemory(bytes - taskBytes);
-        }
-    }
-
-    public void setMemoryReservation(long newMemoryReservation)
-    {
-        checkArgument(newMemoryReservation >= 0, "newMemoryReservation is negative");
-
-        long delta = newMemoryReservation - memoryReservation.get();
-
-        if (delta > 0) {
-            reserveMemory(delta);
-        }
-        else {
-            freeMemory(-delta);
-        }
-    }
-
-    public boolean trySetMemoryReservation(long newMemoryReservation)
+    /**
+     * This method is <b>not</b> thread safe. If multiple threads work on the same
+     * <code>transferredBytesMemoryContext</code> instance it's possible that one of the writes
+     * is lost by the last <code>transferredBytesMemoryContext.setBytes()</code> call.
+     * @param taskBytes The number of bytes to transfer from the OperatorContext to TaskContext.
+     * @param transferredBytesMemoryContext This context should be created with {@link com.facebook.presto.operator.TaskContext#createNewTransferredBytesMemoryContext()}.
+     */
+    public void transferMemoryToTaskContext(long taskBytes, LocalMemoryContext transferredBytesMemoryContext)
     {
-        checkArgument(newMemoryReservation >= 0, "newMemoryReservation is negative");
-
-        long delta = newMemoryReservation - memoryReservation.get();
-
-        if (delta > 0) {
-            if (!driverContext.tryReserveMemory(delta)) {
-                return false;
-            }
-
-            memoryReservation.addAndGet(delta);
+        long bytes = operatorMemoryContext.getUserMemory();
+        operatorMemoryContext.localUserMemoryContext().transferOwnership(transferredBytesMemoryContext);
+        try {
+            transferredBytesMemoryContext.setBytes(transferredBytesMemoryContext.getBytes() + taskBytes - bytes);
```

Comment:
Why not require that `taskBytes <= bytes`? `transferMemoryToTaskContext` should be used for _transferring_ memory, not to make new allocation. (If new allocation is required by the caller, it could bump memory allocation _before_ calling transfer)

Reply:
The current implementation (in master branch) doesn't make this assumption, in fact it handles the case for `taskBytes > bytes` as a separate case: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java#L360